Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,830 @@
+           liftIO $ readIORef ref
+        write_to_session nc' = do
+           ref <- GhcMonad.withSession (return . hsc_NC)
+           liftIO $ writeIORef ref nc'
+
+getModuleExports
+    :: forall m. (GhcMonad m, GmOut m, GmLog m, IOish m)
+    => FilePath
+    -> (FilePath -> [String] -> String -> IO String)
+    -> [GhcPkgDb]
+    -> NiceImportDecl
+    -> m (Maybe ([String], String))
+getModuleExports ghcPkg readProc pkgDbStack m = do
+    let
+        handler :: SomeException -> m (Maybe a)
+        handler e = do
+          let modDoc = doubleQuotes $ text $ modName m
+          gmLog GmDebug "getModuleExports" $
+            (text "Failed to find module" <> modDoc) <+>:
+            text (show e)
+          return Nothing
+
+    minfo <- (findModule (mkModuleName $ modName m) Nothing >>= getModuleInfo)
+               `gcatch` handler
+
+    p <- pkgFindModule ghcPkg readProc pkgDbStack (modName m)
+
+    dflags <- GHC.getSessionDynFlags
+
+    return $ case (minfo, p) of
+        (Nothing,      _) -> Nothing
+        (_,           "") -> Nothing
+        (Just minfo',  _) -> Just (map (showSDocForUser dflags ghcQualify . ppr) $ modInfoExports minfo', p)
+
+-- * e.g. e.g. "base-4.8.2.0:Data.Foldable.length"
+type FullyQualifiedName = String
+
+-- * e.g. "Data.List"
+type StrModuleName      = String
+
+
+data MySymbol = MySymbolSysQualified  String
+              -- ^ e.g. "base-4.8.2.0:Data.Foldable.length"
+              | MySymbolUserQualified String
+                -- ^ e.g. "DL.length" with an import earlier like "import
+                -- qualified Data.List as DL"
+              deriving Show
+
+data ModuleExports = ModuleExports
+    { mName            :: StrModuleName
+    -- ^ e.g. @"Data.List"@
+    , mPackageName     :: String
+    -- ^ e.g. @"snap-0.14.0.6"@
+    , mInfo            :: NiceImportDecl
+    -- ^ Our parse of the module import, with info like "hiding (map)".
+    , qualifiedExports :: [FullyQualifiedName]
+    -- ^ e.g.
+    -- @
+    --        [ "base-4.8.2.0:GHC.Base.++"
+    --        , "base-4.8.2.0:GHC.List.filter"
+    --        , "base-4.8.2.0:GHC.List.zip"
+    --        , ...
+    --        ]
+    -- @
+    } deriving Show
+
+-- User qualified the symbol, so we can filter out anything that doesn't have a
+-- matching 'modImportedAs'.
+refineAs :: MySymbol -> [ModuleExports] -> [ModuleExports]
+refineAs (MySymbolUserQualified userQualSym) exports = filter f exports
+  where
+    f ModuleExports { mInfo = NiceImportDecl { modImportedAs = modAs } } =
+        modAs == moduleOfQualifiedName userQualSym
+
+-- User didn't qualify the symbol, so we have the full system qualified thing,
+-- so do nothing here.
+refineAs (MySymbolSysQualified _) exports = exports
+
+refineRemoveHiding :: [ModuleExports] -> [ModuleExports]
+refineRemoveHiding exports =
+    map (\e -> e { qualifiedExports = removeHiding e }) exports
+  where
+    removeHiding export = filter (`notElem` hiding') thisExports
```

Comment:
`thisExports :: [FullyQualifiedName]` but `hiding' :: [QualifiedName]`, those will never match AFAIU. The type system could've saved you here. Next time use it :p